polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -1451,13 +1451,16 @@ ValueCategory MLIRScanner::VisitCastExpr(CastExpr *E) {
     if (llvmType.isa<LLVM::LLVMPointerType>())
       return ValueCategory(builder.create<mlir::LLVM::NullOp>(loc, llvmType),
                            /*isReference*/ false);
-    else
+    else {
+      auto MT = llvmType.cast<MemRefType>();
```

Comment:
Is it guaranteed that here the cast will succeed?

Reply:
it supposed to, as it is used as the result type of Pointer2Memref.

Reply:
I don't think I follow.  You are visiting a `CastExpr` here.  If you are talking about the implementation that was already there (where they were creating a `Pointer2Memref`, that well can be a limitation of the current implementation.  I would suggest at least add an assert here

Reply:
`cast` should have failed if the expected type is incorrect. I don't mind adding a `llvm_unreachable` to ensure that's the case.

Reply:
> `cast` should have failed if the expected type is incorrect. 

Fail is still a possibility for other examples.  It did not fail for existing test cases does not mean it will always succeed, unless I am missing something.  At least it is non-trivial to me why only Pointer and MemRefType are the only two types possible here.